Profiling support

mgencur opened the pull request:
<!--
Request Prow to automatically lint any go code in this PR:

/lint
-->

Fixes https://github.com/knative/serving/issues/4982

## Proposed Changes

* Introduce a new port 8008 where a web server will be listening and serving profiling data that are consumable by the pprof tool
* Introduce profiling.enable flag in config-observability.yaml that will be either true or false. It will indicate whether it's possible to access the URLs that server pprof-specific profiling data. 
* The profiling data are available for all pods that are currently running in the Knative Serving namespace
* Retrieving profiling data is on demand via wget/curl or web browser accessing certain urls

Steps to get profiling data:
1) Edit the config map and add `profiling.enable = "true"`
```oc edit configmap config-observability -n knative-serving```
2) Use port-forwarding to get access to Knative Serving pods
```
AUTOSCALER_POD=$(oc get pods -n knative-serving | grep autoscaler | grep -v hpa | awk '{print $1}')
kubectl port-forward -n knative-serving $AUTOSCALER_POD  8008:8008
```
3) Use either of these commands (detailed description at [url](https://golang.org/src/net/http/pprof/pprof.go)) to get/analyze pprof data:
```
go tool pprof http://localhost:8008/debug/pprof/heap
go tool pprof http://localhost:8008/debug/pprof/profile?seconds=5
go tool pprof http://localhost:8008/debug/pprof/symbol
```
4) Alternatively, use web browser to access `http://localhost:8008/debug/pprof/` and display the index page with information about available profiling options

There's another PR which needs to be reviewed/integrated first: https://github.com/knative/pkg/pull/562
Once it is integrated I'd update the current PR and reference code in knative.dev/pkg

**Release Note**

<!-- Enter your extended release note in the below block. If the PR requires
additional action from users switching to the new release, include the string
"action required". If no release note is required, write "NONE". -->

```release-note

```

mattmoor-sockpuppet commented on config/config-observability.yaml:
```diff
@@ -99,3 +99,9 @@ data:
     # If metrics.backend-destination is not Stackdriver, this is ignored.
     metrics.allow-stackdriver-custom-metrics: "false"
+
+    # profiling.enable indicates whether it is allowed to retrieve runtime profiling data from 
```
<!--whitespace-bot-->
```suggestion
    # profiling.enable indicates whether it is allowed to retrieve runtime profiling data from
```

markusthoemmes commented on cmd/activator/main.go:
```diff
@@ -241,19 +243,27 @@ func main() {
+		profiling.UpdateProfilingFromConfigMap(profilingHandler, logger),
+	} {
+		configMapWatcher.Watch(metrics.ConfigMapName(), observer)
+	}
```
When I commented on this earlier, I kinda envisioned it to look more like this:

```golang
configMapWatcher.Watch(metrics.ConfigMapName(), func(cm *corev1.ConfigMap) {
	metrics.ConfigMapName(), metrics.UpdateExporterFromConfigMap(component, logger)(cm)
	updateRequestLogFromConfigMap(logger, reqLogHandler)(cm)
})
```

The extra Watch calls might not actually hurt, to me they feel more expensive though. For posterity: I proposed to collapse these multiple Watch statements into one to cut down some repitition.

mgencur replied:
Hmm. This is possible as well. I just don't know which of these solutions is more expensive. The original one registers more functions at the beginning (only once) but then calls them directly (possibly many times when a config map is changed during the lifetime of Knative). Your solution registers only one function at the beginning. And later calls this function indirectly multiple times (indirectly because it has to first call the wrappers which construct the function and then call the function itself). I don't have a strong opinion though.

markusthoemmes replied:
the one I posted has sliiiiightly more allocations (maybe?! don't know maybe it's even compiled away).

The point I'm after is calling `Watch` only once. It likely doesn't matter at all, performancewise. it's mostly for code hygiene.

mgencur replied:
We could also refactor [this](https://github.com/knative/serving/blob/master/vendor/knative.dev/pkg/configmap/manual_watcher.go#L38) to accept `o ...Observer` and then call it like 
```
configMapWatcher.Watch(metrics.ConfigMapName(), metrics.ConfigMapName(), 
        metrics.UpdateExporterFromConfigMap(component, logger),
	updateRequestLogFromConfigMap(logger, reqLogHandler))
```

markusthoemmes replied:
that sounds like an even better idea to me! 👍

markusthoemmes commented on cmd/autoscaler/main.go:
```diff
@@ -100,11 +104,18 @@ func main() {
+		profiling.UpdateProfilingFromConfigMap(profilingHandler, logger),
+	} {
+		cmw.Watch(metrics.ConfigMapName(), observer)
+	}
```
See above.

markusthoemmes commented on cmd/webhook/main.go:
```diff
@@ -82,12 +86,19 @@ func main() {
+		profiling.UpdateProfilingFromConfigMap(profilingHandler, logger),
+	} {
+		configMapWatcher.Watch(metrics.ConfigMapName(), observer)
+	}
```
See above.

markusthoemmes commented on config/activator.yaml:
```diff
@@ -44,12 +44,14 @@ spec:
-        - name: h2c-port
+        - name: h2c
           containerPort: 8013
-        - name: metrics-port
```
These changes seem unnecessary.

mgencur replied:
Yeah. I just wanted to unify the approach across activator, autoscaler, controller, webhook. Only here in activator we use the `-port` suffix. Everywhere else it's omitted. And I think it's actually redundant because this whole section is under "ports" section. WDYT?

markusthoemmes replied:
I agree with that. Mind splitting this out in a separate PR?

mgencur replied:
OK.

markusthoemmes commented on pkg/profiling/config.go:
```diff
@@ -0,0 +1,41 @@
+// as an argument, according to the value in the given ConfigMap
+func UpdateProfilingFromConfigMap(profilingHandler *Handler, logger *zap.SugaredLogger) func(configMap *corev1.ConfigMap) {
+	return func(configMap *corev1.ConfigMap) {
+		if profiling, ok := configMap.Data[profilingKey]; ok {
```
Cut down on indentation a bit here by returning early, for example:

```golang
profiling, ok := configMap.Data[profilingKey]
if !ok {
  return
}
enabled, err := strconv.ParseBool(profiling)
if err != nil {
  logger.Errorw("Failed to update profiling", zap.Error(err))
  return
}
logger.Infof("Profiling enabled: %t", enabled)
profilingHandler.Enabled = enabled
```

That makes the happy path unindented which is usually the case in functions.

markusthoemmes commented on pkg/profiling/server.go:
```diff
@@ -0,0 +1,68 @@
+
+// NewHandler create a new ProfilingHandler which serves runtime profiling data
+// according to the given context path
+func NewHandler(profilingEnabled bool) *Handler {
```
`profilingEnabled` seems to be always set to `false`. Should we just ditch the parameter?

mgencur commented:
I suppose the build failure is expected because I have some new stuff in the vendor folder. Once the related PR is merged I'll update this PR and fix the build error by running `./hack/update-codegen.sh`

mattmoor-sockpuppet commented on pkg/autoscaler/statserver/server_testing_test.go:
```diff
@@ -35,6 +35,7 @@ func NewTestServer(statsCh chan<- *autoscaler.StatMessage) *TestServer {
 		Server:     New(testAddress, statsCh, zap.NewNop().Sugar()),
 		listenAddr: make(chan string, 1),
 	}
+	
```
<!--whitespace-bot-->
```suggestion

```

mgencur commented:
Alright, the queu-proxy requires a little more work...

mgencur commented:
I've rebased this. Updated the dependency on knative.dev/pkg and for now removed the queue-proxy support. It might be good to implement this in a separate PR after some discussion.

So this PR is ready for review.

markusthoemmes commented on config/activator.yaml:
```diff
@@ -44,12 +44,14 @@ spec:
         image: knative.dev/serving/cmd/activator
         ports:
-        - name: http1
+        - name: http1-port
```
This seems like a merge failure.

mgencur replied:
Right. Thanks. Fixed.

vagababov commented:
/assign mattmoor

vagababov commented:
Matt, I can't approve configmaps.
/assign mattmoor
